### DISCUSSION TITLE
Update graph module (Dashticz 3.12.0.1)

### DISCUSSION THREAD

**Discussion by lokonli** [Dashticz/dashticz#1172] [General]
Work in progress.

To test, checkout the graphtest branch

```
   git fetch origin
   git checkout graphtest
```

Prelimininary documentation can be found here:
https://dashticz.readthedocs.io/en/graphtest/releasenotes/upgrading/v3.12.1.html

To go back to beta:

```
   git checkout beta
```

Feedback is welcome.

**Comment by JRB202** [comment]
Hi,

I had a first test on new graph and have problems with this simple example :
DeviceIdx = 275;   // which is an electric meter
blocks['Battery_graph'] = {
                                                type: 'graph',
                                                devices: [DeviceIdx],
                                                                        datasets: {
                                                                              v: {
                                                                                borderWidth: 10,
                                                                                legend: 'InWatts',
                                                                                borderDash : [10,10],
                                                                                pointRadius: 5,
                                                                              }  // end v
                                                                         }  // end dataset
                                           } // end block

1) The graph is well rendered with 2 lines : v and eu values lines and legend as v_275 and eu_275 
but none of the dataset defined properties are executed.
==>Where am I wrong with syntax ?

2) I would like to graph as line only the v values and suppress the eu values line
How can I select v only (as v_1 will be taken as a 2nd dataset if I represent 2 devices ...) ?

**Reply by lokonli** [reply]
o yes.
For some device type I create default graph cnfigurations, including energy devices.

To disable the defaults add:
usedefaults: false

You should then get two datasets: 'v' and 'eu'.
To hide 'eu'  you can use one of the following methods.
add:
  graphTypes: 'v' 

This will only show the 'v' datasets.

or add:
datasets: {
   eu: {
      hideData: true
   }
}

I'll double check tomorrow.

**Reply by lokonli** [reply]
Please check you are using the correct branch
(Dashticz->settings->info should show 3.12.0.2)

**Comment by lokonli** [comment]
![image](https://github.com/Dashticz/dashticz/assets/22852111/c38857bf-a989-487a-a1d3-24ae4d347a25)

blocks['plug'] = {
    type: 'graph',
    devices: [1343],
    graphTypes: [ 'v'],
    usedefaults: false,
    datasets: {
        v: {
        borderWidth: 10,
        legend: 'InWatts',
        borderDash : [10,10],
        pointRadius: 5,
        } // end v
        } // end dataset
}

**Comment by JRB202** [comment]
Thanks for reply.

Super, I just added the line :
graphTypes: [ 'v'], usedefaults: false,

and now dataset properties are well executed for 'v' only.

1) You should add this line to your example in the doc here :
https://dashticz.readthedocs.io/en/graphtest/releasenotes/upgrading/v3.12.1.html#new-block-parameter-datasets

2) Suggestion : I noticed that borderDash is also rendered in legend with a really bad look 
especially if a large borderWidth is also used ...

**Reply by lokonli** [reply]
For point 2: fixed with 3.12.0.3

**Comment by JRB202** [comment]
I left with a few non working syntax in datasets definition :

1) borderColor: 'yellow' or datasetColors: 'yellow', are not working to color lines when used in datasets new block
the only way I found is :
datasetColors: ['green','blue','red','yellow'],
used at graph block definition level

2) when using legend: 'Production',  in Dataset definition it is well rendered
But the default legends with v  v_1 v_2 are also displayed
I tried legend: false, at graph definition level but no more legends at all.
==> the default v  v_1 v_2 legends must be suppressed if legend is defined at dataset level
and the colors associated with these unwanted  v  v_1 v_2 legends are overwriting colors defined in 
datasetColors: ['green','blue','red','yellow'], defined at graph level
==> Colors defined at dataset level must overwrite colors defined at graph level or defined by default in your code.

**Reply by lokonli** [reply]
For 1: borderColor fixed in 3.12.0.4
For the fill color (not the border color) you can use the color parameter

For 2:
You can add hideData: true and hidefromlegend:true for the datasets you don't want to show.

I'll consider to use the legend parameter again to show certain datasets only.
I will check color behavior.

**Comment by JRB202** [comment]
Other fix is needed here :
the Value displayer which is hover activated on a displayed graph contains only default legend (v v_1 v_2) and associated values 
but the legend and values as defined (with formulas) in the dataset are not displayed at all

**Comment by JRB202** [comment]
Errors in Console :
I see an error in console of type 404 on this url :
https://raw.githubusercontent.com/Dashticz/dashticz/test/version.txt

and hundreds of messages from :
sw-toolbox.js:15 
when reading successfully domoticz log with requests like :
/json.htm?type=command&param=getlog&lastlogtime=1708712963&loglevel=268435455

**Reply by lokonli** [reply]
You can ignore the 404 error. (The test branch doesn't exist)

sw-toolbox.js is not used by Dashticz, as far as I know. Can you find the callstack of the command that generates the warning? What is the last file+linenumber from the Dashticz source code in the call stack?

**Comment by JRB202** [comment]
After upgrade to V3.12.0.4 test
New problem with  formulas which are not executed in custom reports :
v1, v_1 and v_2 but not  'v-v_1-v_2' neither  'd.v-d.v_1-d.v_2'
MyDatasets = { 
                                                         'v': { color : 'green' },
                                                         'v_1': { color : 'blue' },
                                                         'v'_2: { color : 'red' },
                                                        'v-v_1-v_2': { color : 'yellow' },
}

**Reply by lokonli** [reply]
The 'v-v_1-v_2' notation is not supported.
You still have to define it as a custom graph.

like:
```
blocks['graph_43'] = {
    devices: [43],
    custom: {
        lasthours: {
            range: 'day',
            filter: '4 hours',
            data: {
                nett: 'd.v+d.v2-d.r1-d.r2',
                usage: 'd.v+d.v2',
                generation: '-d.r1-d.r2'
            }
          },//lasthours
      }, //custom
       datasets: {
             nett: {color: ' green'},
             usage: {color: 'blue'},
             generation: { color: 'yellow' }
       }
```

Probably I'll add the data part of the custom property to dataset,like: 
```
blocks['graph_43'] = {
    devices: [43],
    custom: {
        lasthours: {
            range: 'day',
            filter: '4 hours',
        },
     },
     datasets: {
             nett: {color: ' green', data: 'd.v+d.v2-d.r1-d.r2'},
             usage: {color: 'blue', data: 'd.v+d.v2'},
             generation: { color: 'yellow', data:  generation: '-d.r1-d.r2'}
       }
```
and rename 'custom' to 'buttons' 
(this last example is not implemented yet)

**Comment by JRB202** [comment]
This structure is clearly the good solution as the "old" custom way do not support dataset level properties

datasets: {
             nett: {color: ' green', data: 'd.v+d.v2-d.r1-d.r2'},
             usage: {color: 'blue', data: 'd.v+d.v2'},
             generation: { color: 'yellow', data:  generation: '-d.r1-d.r2'}
       }

Tell me when implemented, I will be happy to test

**Reply by lokonli** [reply]
implemented.

Check instructions on:
 https://dashticz.readthedocs.io/en/graphtest/releasenotes/upgrading/v3.12.1.html#data-parameter-in-datasets-section

**Comment by JRB202** [comment]
Hi Lokonli
I tested the new dataset version and found a problem :
The custom version of this example : 
https://dashticz.readthedocs.io/en/graphtest/releasenotes/upgrading/v3.12.1.html#data-parameter-in-datasets-section

displays formula based datasets perfectly BUT :
also displays v, v_1 and v_2 in data graph and legend on top of custom formula based datasets defined 

This is unwanted behavior as we are supposed to see only the formula based datasets

**Reply by lokonli** [reply]
Try adding the following to the block definition:

showAllDatasets: false,

**Comment by JRB202** [comment]
showAllDatasets: false, is masking all the expected datasets lines and bars except the 1rst expected dataset.
but the result on legend is correct with this option and showing only expected datasets
I tried with 2 different custom graphs.
Something is clearly wrong with the display of bar and lines
